File: ts/512/436.ts
function findRightInterval(intervals: number[][]): number[] {
  const n = intervals.length,
        ht = new Map<number, number>(),
        ans: number[] = new Array(n).fill(-1)
  const binary_search = (intervals: number[][], st: number, target: number): number => {
    let l = st, r = intervals.length - 1, m = 0
    if (intervals[r][0] < target) return -1
    while (l < r) {
      m = l + ((r - l) >> 1)
      if (intervals[m][0] >= target) r = m
      else l = m + 1
    }
    return r
  }
  for (let i = 0; i < n; ++i) {
    ht.set(intervals[i][0], i)
  }
  intervals.sort((a, b): number => a[0] - b[0])
  for (let i = 0; i < n; ++i) {
    const idx = binary_search(intervals, i, intervals[i][1])
    if (idx >= 0) ans[ht.get(intervals[i][0])] = ht.get(intervals[idx][0])
  }
  return ans
};

File: py3/512/436.py
from typing import List

class Solution:
    def findRightInterval(self, intervals: List[List[int]]) -> List[int]:
        def sort_key(nums):
            return nums[0]
        def search_index(intervals, x):
            l, r = 0, len(intervals) - 1
            if intervals[r][0] < x:
                return -1
            while l < r:
                m = l + (r - l) // 2
                if intervals[m][0] == x: # 每个区间的 start 都是唯一的，当 start == end，一定是最小的
                    r = m
                    break
                if intervals[m][0] < x:
                    l = m + 1
                else:
                    r = m
            return r

        ht = {v[0]: i for i, v in enumerate(intervals)}
        ans = [-1] * len(intervals)
        intervals.sort(key=sort_key)

        for v in intervals:
            idx = search_index(intervals, v[1])
            if idx >= 0:
                ans[ht[v[0]]] = ht[intervals[idx][0]]

        return ans